packages/datacommons-mcp/evals/test_tool_agent/evals/test_tools.py

```diff
@@ -1,18 +1,244 @@
+        if "overall_eval_status" in result_df.columns:
+            failed_rows = result_df[result_df["overall_eval_status"] == "FAILED"]
+            if not failed_rows.empty:
+                # Count failures by metric type
+                failure_summary = []
+
+                if "tool_eval_status" in result_df.columns:
+                    tool_failures = len(
+                        failed_rows[failed_rows["tool_eval_status"] == "FAILED"]
+                    )
+                    if tool_failures > 0:
+                        failure_summary.append(f"tool_eval: {tool_failures} failures")
+
+                if "response_eval_status" in result_df.columns:
+                    response_failures = len(
+                        failed_rows[failed_rows["response_eval_status"] == "FAILED"]
+                    )
+                    if response_failures > 0:
+                        failure_summary.append(
+                            f"response_eval: {response_failures} failures"
+                        )
+
+                pytest.fail(
+                    f"Evaluation FAILED for {path.name}.\n"
+                    f"Failed metrics summary: {', '.join(failure_summary) if failure_summary else 'unknown failures'}\n"
+                    f"Total failed evaluations: {len(failed_rows)}"
+                )
+
+    @classmethod
+    def teardown_class(cls) -> None:
+        """
+        Pytest hook that runs once after all test methods in the class are done.
+        This is the perfect place to generate the final report.
+        """
+        if not cls.all_results_dfs:
+            print("\nNo DataFrames were collected to generate a report.")
+            return
+
+        print(
+            f"\n📈 Combining {len(cls.all_results_dfs)} result(s) into a single report..."
+        )
+
+        # Combine all collected DataFrames into one large DataFrame
+        final_report_df = pd.concat(cls.all_results_dfs, ignore_index=True)
+
+        # Write the final DataFrame to a single, easy-to-read HTML file
+        try:
+            # Make report output directory if it doesn't exist
+            pathlib.Path(REPORT_OUTPUT_DIR).mkdir(parents=True, exist_ok=True)
+            # Write output path with a timestamp
+            report_output_path = (
+                pathlib.Path(REPORT_OUTPUT_DIR)
+                / f"{REPORT_OUTPUT_BASE_FILENAME}-{pd.Timestamp.now().strftime('%Y%m%d-%H%M%S')}.html"
+            )
+            create_styled_html_report(final_report_df, report_output_path)
+            print(f"✅ Report successfully generated at: {report_output_path}")
+            # Write to CSV as well for easier data manipulation
+            csv_path = report_output_path.with_suffix(".csv")
+            final_report_df.to_csv(csv_path, index=False)
+            print(f"✅ CSV report generated at: {csv_path}")
+        except Exception as e:
+            print(f"🔥 Failed to write HTML report: {e}")
+
+
+def create_styled_html_report(
```

Comment:
Would it make sense to move the report generation to be a static AgentEvaluator method?